pkg/expr/expr.go

```diff
@@ -389,17 +390,43 @@ func Parse(src string) (ex *Expression, err error) {
 	return ex, nil
 }
 
+// DefaultCacheSize is the default size for the expression cache.
+const DefaultCacheSize = 1024
+
 // Evaluator for a c-like expression language.
 type cexl struct {
-	//TODO add ast cache
+	cache *lru.Cache
 	// function Map
 	fMap map[string]FuncBase
 }
 
+func (e *cexl) cacheGetExpression(exprStr string) (ex *Expression, err error) {
+
+	// TODO: add normalization for exprStr string, so that 'a | b' is same as 'a|b', and  'a == b' is same as 'b == a'
+
+	if v, found := e.cache.Get(exprStr); found {
+		return v.(*Expression), nil
+	}
+
+	if glog.V(4) {
+		glog.Infof("expression cache miss for '%s'", exprStr)
+	}
+
+	ex, err = Parse(exprStr)
+	if err != nil {
+		return nil, err
+	}
+	if glog.V(4) {
+		glog.Infof("caching expression for '%s''", exprStr)
+	}
+
+	_ = e.cache.Add(exprStr, ex)
+	return ex, nil
+}
+
 func (e *cexl) Eval(s string, attrs attribute.Bag) (ret interface{}, err error) {
 	var ex *Expression
-	// TODO check ast cache
-	if ex, err = Parse(s); err != nil {
+	if ex, err = e.cacheGetExpression(s); err != nil {
```

Comment:
Do we have a cache somewhere to avoid evaluating the expression if we've already done so with the same attributes? If not, is there a github issue for that?